Upload consensus genomes to id3c

kairstenfay opened the pull request:
Add steps to the Snakefile to convert Fasta & summary stats files to Json and then upload to ID3C.
See seattleflu/id3c/pull/48 for the ID3C changes.

I am getting seemingly unrelated errors when running the Snakefile locally. The upload to ID3C step runs, but steps down the line don't. Could someone test these changes locally? I don't think my edits changed anything but I need additional confirmation (and to fix my assembly setup, apparently).

kairstenfay committed:
Trim trailing whitespace

kairstenfay commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Upload consensus genomes to ID3C after assembly](https://trello.com/c/kBEezty1/3-upload-consensus-genomes-to-id3c-after-assembly)

joverlee521 commented on Snakefile:
```diff
@@ -469,6 +472,71 @@ rule fasta_headers:
+
+rule summary_stats_to_json:
+    input:
+        bam_coverage = rules.coverage_summary.output.coverage,
```
bamstats summary info should be coming from `rules.bamstats.output.bamstats_file`

joverlee521 commented:
I didn't realize there was a `consensus-genome-uploader` role that I needed to assign to the test user, so the upload to receiving failed, but this was not reflected anywhere in the Snakemake output. Should we have some way for the pipeline to catch that the upload had failed and then try again?

kairstenfay commented:
> I didn't realize there was a `consensus-genome-uploader` role that I needed to assign to the test user, so the upload to receiving failed, but this was not reflected anywhere in the Snakemake output. Should we have some way for the pipeline to catch that the upload had failed and then try again?

Ah, I should've made that clearer.
And yes, that would be excellent! Right now I have the upload step printing to an `http-response.log`, but I would rather give the user feedback & interrupt the Snakefile. I just haven't gone digging in the Snakemake documentation yet to find how to do that. Great idea.

kairstenfay commented:
@joverlee521 I made a commit just now to fix the source of the bamstats file. Also, it includes some Snakefile improvements I made after reading the Snakemake docs. 

Right now an error shows in the DAG if the server returns no response from `curl` (essentially if there's something wrong with the server). However, any other response other than 200 (401 Unauthorized, etc.) doesn't return an error but does show up in the log provided. 

Do you have any idea on how to raise an error from the log? Or maybe you can think of a better way to approach this?

joverlee521 commented:
Looks like you're already trying with checkpoint? This would be my way to do this: write a `checkpoint check_http_response` after the post request that looks at the log and sees what response code was given. I'm not sure how to make it error out depending on the response. That will need some investigation.

kairstenfay commented:
Yes I gave `checkpoint` a try, but it only is useful for dependencies of
the checkpoint. I use a checkpoint when constructing the payload, but no
checkpoint for the POST rule because nothing depends on it.

I'm not sure how to trigger an error for a bad HTTP response, either. This
may require some digging around in bash.

On Wed, Jul 24, 2019, 9:14 AM Jover Lee <notifications@github.com> wrote:

> Looks like you're already trying with checkpoint? This would be my way to
> do this: write a checkpoint check_http_response that looks at the log and
> sees what response code was given. I'm not sure how to make it error out
> depending on the response. That will need some investigation.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/seattleflu/assembly/pull/7?email_source=notifications&email_token=ACW3NEBKIRNYGRZA7DELQD3QBB5VHA5CNFSM4IGHE5J2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD2W27CY#issuecomment-514699147>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACW3NEBFZMGNH5VOWYQ4WW3QBB5VHANCNFSM4IGHE5JQ>
> .
>

kairstenfay commented:
Slack alerts are now a part of the assembly pipeline for failed ID3C uploads :tada: 

Note: there is no Slack alert sent currently if the `curl` command gets no response from the server (i.e. if the server is not running). However, Snakemake will show the job as failed.
For bad response codes (other than 204), Slack alerts are sent.

kairstenfay commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Create Slack alerts for assembly jobs that failed to upload to ID3C](https://trello.com/c/6NUZawDg/176-create-slack-alerts-for-assembly-jobs-that-failed-to-upload-to-id3c)

kairstenfay committed:
Add steps to upload consensus genomes to ID3C

Add five steps to the Snakefile:
1. an intermediate step to create a Json file containing metadata like
URLs
2. an intermediate step to convert the masked consensus fasta file to
Json format
3. an intermediate step to convert the bowtie2 and BAMstats summary
statistics to Json format
4. an intermediate checkpoint to combine the Json files from steps 1, 2,
and 3 to create the final ID3C payload.
5. a final step to upload the newly created payload from step 4 to ID3C
via a HTTP POST request to a `/receiving/consensus-genome` endpoint. The
connection details including specific URL and username/password are
stored in the environment.

If step 5 receives a non-successful HTTP response, then the
response from the URL is stored in a log under the given wildcard
directory/file path. The job does not appear as failed in the Snakemake
DAG.

Co-authored-by: Jover Lee <joverlee521@gmail.com>

tsibley commented on scripts/setup_config_file.py:
```diff
@@ -186,14 +186,11 @@ def find_read_ids(fastq_file: str) -> set:
+    edit_config_file(args.directory, args.config_file,
+                        args.sample_target_map,
+                        args.sample, args.target,
+                        args.max_difference)
```
This diff hunk includes more than just whitespace changes.

tsibley commented on Snakefile:
```diff
@@ -443,7 +443,7 @@ rule nwgc_sfs_map:
     shell:
         """
-        python3 scripts/id_barcode_key_value.py \
+        python scripts/id_barcode_key_value.py \
```
I'm surprised this is necessary. I thought conda installed both commands. Weird! My conda environments contain both. Maybe a version thing with conda?

<details><summary>My <code>conda info</code> :point_down: </summary>

```
     active environment : None
            shell level : 0
       user config file : /home/tom/.condarc
 populated config files : /home/tom/.condarc
          conda version : 4.6.14
    conda-build version : not installed
         python version : 3.7.3.final.0
       base environment : /home/tom/.conda  (writable)
           channel URLs : https://conda.anaconda.org/bioconda/linux-64
                          https://conda.anaconda.org/bioconda/noarch
                          https://conda.anaconda.org/conda-forge/linux-64
                          https://conda.anaconda.org/conda-forge/noarch
                          https://repo.anaconda.com/pkgs/main/linux-64
                          https://repo.anaconda.com/pkgs/main/noarch
                          https://repo.anaconda.com/pkgs/free/linux-64
                          https://repo.anaconda.com/pkgs/free/noarch
                          https://repo.anaconda.com/pkgs/r/linux-64
                          https://repo.anaconda.com/pkgs/r/noarch
          package cache : /home/tom/.conda/pkgs
       envs directories : /home/tom/.conda/envs
               platform : linux-64
             user-agent : conda/4.6.14 requests/2.21.0 CPython/3.7.3 Linux/4.15.0-54-generic ubuntu/18.04.2 glibc/2.27
                UID:GID : 1000:1000
             netrc file : None
           offline mode : False
```

</details>

kairstenfay replied:
Yep, you're right. I updated my conda and now `python3` is installed.

tsibley commented on Snakefile:
```diff
@@ -469,6 +472,81 @@ rule fasta_headers:
+        temp("consensus_genomes/{reference}/{sample}.metadata.json")
+    shell:
+        """
+        python scripts/metadata_to_json.py "{input.all_r1}" "{input.all_r2}" > {output}
```
It's more robust to use `{input.all_r1:q}` instead of `"{input.all_r1}"` because it's still possible for shell metachars to leak in the latter. The `:q` is a Snakemake formatting code which appropriately **q**uotes the value for the shell. More info [in the docs](https://snakemake.readthedocs.io/en/stable/snakefiles/rules.html?highlight=%3Aq#rules).

(In general, I prefer to use `:q` for almost all interpolations in Snakemake, but I'm in the minority in the lab.)

kairstenfay replied:
I'm glad you pointed that out. Sending multiple arguments as a string was why the string had to be split in the `metadata_to_json` script. The code is improved using `:q`

tsibley commented on Snakefile:
```diff
@@ -469,6 +472,81 @@ rule fasta_headers:
+            --bowtie2 {input.bowtie2} > {output}
+        """
+
+checkpoint create_id3c_payload:
```
It likely doesn't matter, but I'm not sure this strictly needs to be a checkpoint rule?

tsibley commented on Snakefile:
```diff
@@ -469,6 +472,81 @@ rule fasta_headers:
+        rules.create_id3c_payload.output
+    params:
+        id3c_url = config['id3c-consensus-genome-post-url'],
+        id3c_username_and_password = config['id3c-username-and-password'],
```
I would prefer the API URL and credentials to come from the environment rather than a file. This will make it easier to deploy later.

tsibley commented on Snakefile:
```diff
@@ -469,6 +472,81 @@ rule fasta_headers:
+        then
+            echo "Something went wrong in the ID3C POST request.\nSee {log} for more information."
+            exit $response
+        fi
```
Since we're in a Python environment already, this might be a little cleaner if done using a Snakemake `run` block and the `requests` package.

If we still with shell, then the username/password should definitely get a `:q` modifier on the interpolation, as the password may contain shell metachars.

tsibley commented on scripts/metadata_to_json.py:
```diff
@@ -0,0 +1,34 @@
+
+    print(json.dumps( {
+        "metadata": {
+            "URLs": list(set(urls))
```
I think I'd rather consistently use lowercase field names.

tsibley commented on scripts/metadata_to_json.py:
```diff
@@ -0,0 +1,34 @@
+    parser.add_argument("urls_r1", type=str, nargs="?",
+        metavar = "<urls_r1>")
+    parser.add_argument("urls_r2", type=str, nargs="?",
+        metavar = "<urls_r2>")
```
nargs can be set differently to take multiple values, which would avoid the need to split on spaces.

tsibley commented on Snakefile:
```diff
@@ -532,7 +533,9 @@ rule post_masked_consensus_and_summary_stats_to_id3c:
         id3c_username_and_password = config['id3c-username-and-password'],
-        id3c_expected_response_code = 204
+        id3c_expected_response_code = 204,
+        id3c_slack_webhook = config['id3c-alerts-slack-webhook'],
```
Ditto this webhook URL.

tsibley commented on Snakefile:
```diff
@@ -545,6 +548,11 @@ rule post_masked_consensus_and_summary_stats_to_id3c:
             echo "Something went wrong in the ID3C POST request.\nSee {log} for more information."
+            curl -X POST -H 'Content-type: application/json' \
+                --data '{{"text": \
+                ":rotating_light: @{params.user} Assembly failed to upload to ID3C.\nMore details at `{log}`"}}' \
```
Rather than constructing this JSON by hand, it'd be more robust to do something like:

```
params:
    slack_payload = {
        "text": f":rotating_light: @{params.user} Assembly failed to upload to ID3C.\nMore details at `{log}`"
    }
shell:
    … --data {json.dumps(params.slack_payload):q} …
```

kairstenfay replied:
This is a good suggestion, but I am abandoning curl in favor of python requests.

tsibley reviewed (changes requested):
Exciting to see this come together! Thanks both for your work on it while I was away.

I left a few notes with things I'd like to see changed. Happy to chat about it more too, or dive in myself since @kairstenfay is out for a bit longer.

kairstenfay commented:
@tsibley I am on board with all these requested changes. Feel free to work them in if you want to get this merged with master sooner. Otherwise I'll work on it when I get back. Thanks!

tsibley commented:
@kairstenfay Sounds good!

kairstenfay committed:
Remove try/catch block that prevents errors

Remove the try/catch block from the setup config file script. It was
preventing informative error messages from printing to the console.

kairstenfay committed:
Create Slack alerts for failed assembly uploads

If a POST request to the ID3C consensus genome web API endpoint returns
a non-successful HTTP response during an assembly job, create an alert
in the #id3c-alerts Slack channel notifying the user of their failed
job.

This feature is necessary because the Snakemake DAG does not show failed
POST requests as failed jobs in the DAG. Therefore, the current solution
is to bring attention to failed uploads via Slack.

The Slack webhook is stored in the environment.

Co-authored-by: Jover Lee <joverlee521@gmail.com>

tsibley commented on Snakefile:
```diff
@@ -469,8 +472,94 @@ rule fasta_headers:
+    params:
+        id3c_url = os.environ['ID3C_ASSEMBLY_ENDPOINT'],
+        id3c_username = os.environ['ASSEMBLY_ID3C_USERNAME'],
+        id3c_password = os.environ['ASSEMBLY_ID3C_PASSWORD']
```
It would be good to have these environment variables consistently named, maybe for example:

`ID3C_URL`
`ID3C_USERNAME`
`ID3C_PASSWORD`

tsibley commented on Snakefile:
```diff
@@ -469,8 +472,94 @@ rule fasta_headers:
+
+        except Exception as err:
+            file.write(str(err))
+            raise Exception(f"Error: {r.status_code} in ID3C POST request. See {log}")
```
`r` is not defined here?

tsibley commented on Snakefile:
```diff
@@ -469,8 +472,94 @@ rule fasta_headers:
+
+        try:
+            response = requests.post(
+                params.id3c_url,
```
I think it will be better to do `urljoin(params.id3c_url, "/v1/receiving/consensus-genome")` instead of expecting the full endpoint in the environment variable. The code here in the Snakefile is written to the expectations of that particular `/v1` endpoint and so it's good to explicitly define that here rather than make it implicit.

tsibley commented on Snakefile:
```diff
@@ -469,8 +472,94 @@ rule fasta_headers:
+        file = open(str(log), "w")
+
+        with open(str(input)) as f:
+            data = json.load(f)
```
No need to do `json.load` and then `json.dumps`. Instead, we can read the file as a string/bytes and pass it along unparsed.

tsibley commented on Snakefile:
```diff
@@ -532,7 +534,8 @@ rule post_masked_consensus_and_summary_stats_to_id3c:
         id3c_username = os.environ['ASSEMBLY_ID3C_USERNAME'],
-        id3c_password = os.environ['ASSEMBLY_ID3C_PASSWORD']
+        id3c_password = os.environ['ASSEMBLY_ID3C_PASSWORD'],
+        id3c_slack_webhook = os.environ['ID3C_SLACK_WEBHOOK_URL'],
```
Would probably just call this `SLACK_WEBHOOK_URL` since it's not an ID3C thing?

kairstenfay replied:
It was the #id3c channel, but I agree it's an unnecessarily long variable name

tsibley commented on Snakefile:
```diff
@@ -553,8 +556,22 @@ rule post_masked_consensus_and_summary_stats_to_id3c:
             file.write(str(http_err))
 
+            slack_data = { "text":
+                f":rotating_light: @{getpass.getuser()} Assembly failed to upload to ID3C.\nMore details at `{log}`"
```
The local user won't always match the Slack user, for what that's worth. Nice to have when it does, though.

tsibley replied:
Could the Slack message include some bit of the HTTP error? That might help fasting troubleshooting.

kairstenfay replied:
It turns out I was overly optimistic; Slack mentions through the API actually require the User ID (not the username) which requires a separate API lookup through Slack. 

I'm going to abandon this idea for now and just call out the user by their local username on their computer.

tsibley reviewed (changes requested):
Thanks for addressing the previous comments! I have a few more small things.

One question I had is about this ¶ in the last commit message:

> This feature is necessary because the Snakemake DAG does not show failed
> POST requests as failed jobs in the DAG. Therefore, the current solution
> is to bring attention to failed uploads via Slack.

It seems like the reason the DAG doesn't have a failed job is because the job code suppresses errors. If a failed job (in addition to a Slack notification!) would be a better outcome if the upload fails, we can make it happen. @joverlee521 your thoughts here would be appreciated too.

tsibley commented on Snakefile:
```diff
@@ -46,8 +46,10 @@ https://github.com/lmoncla/illumina_pipeline
+import getpass
 import requests
 from requests.exceptions import HTTPError
+import getpass
```
`getpass` is imported twice

kairstenfay commented on Snakefile:
```diff
@@ -575,10 +576,12 @@ rule post_masked_consensus_and_summary_stats_to_id3c:
                 file.write(str(http_err))
 
+            raise http_err from None
+
```
Thanks for your work in this commit, @joverlee521 ! I have everything rebased and ready to re-push. I'm just curious about this line here

`raise http_err from None`

What made you want to add that?

joverlee521 replied:
> It seems like the reason the DAG doesn't have a failed job is because the job code suppresses errors. If a failed job (in addition to a Slack notification!) would be a better outcome if the upload fails, we can make it happen.

This was a response to @tsibley comment earlier. I added that to cause the pipeline to completely error out if there is an error with the upload. Not sure if it's necessary?

kairstenfay replied:
Oh, that makes sense! I'll do some more testing with it and re-push soon.

kairstenfay replied:
Neat fix. How does it work? 

Does an error only get raised if `http_err` is defined? and why do we ned `from None`? This is cool and I want to understand it better.

joverlee521 replied:
I honestly just copied the `raise _ from None` from our ID3C code. It might not be the right syntax based on this [article](https://codecalamity.com/exception-exception-read-all-about-it/#from-what)?

tsibley replied:
The `from None` clause is not doing anything in this case because the `raise` isn't happening in a nested exception handler: it's in the first `except` for the POST to ID3C, not the second `except` for the POST to Slack. If it was in the latter, the `from None` would suppress reporting of the ID3C exception and just report the Slack exception. I've used this from in various places in ID3C when I want to convert from exception A to exception B without reporting extra confusing details about A.

Additionally, there's also some variable shadowing going on because both `except` blocks use the same variable name for the exception. This means that by the time the `raise http_err from None` line is reached, `http_err` is actually undefined, which produces a NameError. The NameError still causes the build to fail, but not for the reason we wanted! :-) and not with the details of the HTTP exception. This can be fixed by using different variable names for the two exceptions.

kairstenfay replied:
Fixed it! Thanks, Thomas.

tsibley reviewed (changes requested):
I haven't tried this out, but I trust you have, and the code looks good. Thank you both! Would be good to fix the exception thing I responded to your comments about, but otherwise this looks good for merge.
